### PR TITLE
ImageConfig should contain the image from sandbox.config

### PR DESCRIPTION
**Pull request by pingsutw**
Signed-off-by: Kevin Su <pingsutw@apache.org>

# TL;DR
Failed to run this [example](https://github.com/flyteorg/flytesnacks/blob/b86f823f4e6c2100615d5fe3b0ceb336a042911e/cookbook/core/containerization/multi_images.py#L77) because `ImageConfig` doesn't contain the image from [sandbox.config](https://github.com/flyteorg/flytesnacks/blob/b86f823f4e6c2100615d5fe3b0ceb336a042911e/cookbook/core/sandbox.config#L4-L6)

```bash
(flytekit-3.9) ➜  core git:(master) ✗ pyflyte --config sandbox.config run --remote containerization/multi_images.py my_workflow
Traceback (most recent call last):
  File "/Users/kevin/opt/anaconda3/envs/flytekit-3.9/bin/pyflyte", line 33, in <module>
    sys.exit(load_entry_point('flytekit', 'console_scripts', 'pyflyte')())
  File "/Users/kevin/opt/anaconda3/envs/flytekit-3.9/lib/python3.9/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/Users/kevin/opt/anaconda3/envs/flytekit-3.9/lib/python3.9/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/Users/kevin/opt/anaconda3/envs/flytekit-3.9/lib/python3.9/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/kevin/opt/anaconda3/envs/flytekit-3.9/lib/python3.9/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/kevin/opt/anaconda3/envs/flytekit-3.9/lib/python3.9/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/kevin/opt/anaconda3/envs/flytekit-3.9/lib/python3.9/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/kevin/opt/anaconda3/envs/flytekit-3.9/lib/python3.9/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/Users/kevin/git/flytekit/flytekit/clis/sdk_in_container/run.py", line 519, in _run
    remote_entity = remote.register_script(
  File "/Users/kevin/git/flytekit/flytekit/remote/remote.py", line 625, in register_script
    return self.register_workflow(entity, serialization_settings, version, default_launch_plan, options)
  File "/Users/kevin/git/flytekit/flytekit/remote/remote.py", line 499, in register_workflow
    ident = self._serialize_and_register(entity, serialization_settings, version, options)
  File "/Users/kevin/git/flytekit/flytekit/remote/remote.py", line 391, in _serialize_and_register
    _ = get_serializable(m, settings=serialization_settings, entity=entity, options=options)
  File "/Users/kevin/git/flytekit/flytekit/tools/translator.py", line 581, in get_serializable
    cp_entity = get_serializable_workflow(entity_mapping, settings, entity, options)
  File "/Users/kevin/git/flytekit/flytekit/tools/translator.py", line 226, in get_serializable_workflow
    upstream_node_models = [
  File "/Users/kevin/git/flytekit/flytekit/tools/translator.py", line 227, in <listcomp>
    get_serializable(entity_mapping, settings, n, options)
  File "/Users/kevin/git/flytekit/flytekit/tools/translator.py", line 584, in get_serializable
    cp_entity = get_serializable_node(entity_mapping, settings, entity, options)
  File "/Users/kevin/git/flytekit/flytekit/tools/translator.py", line 412, in get_serializable_node
    task_spec = get_serializable(entity_mapping, settings, entity.flyte_entity, options=options)
  File "/Users/kevin/git/flytekit/flytekit/tools/translator.py", line 578, in get_serializable
    cp_entity = get_serializable_task(entity_mapping, settings, entity)
  File "/Users/kevin/git/flytekit/flytekit/tools/translator.py", line 173, in get_serializable_task
    container = entity.get_container(settings)
  File "/Users/kevin/git/flytekit/flytekit/core/python_auto_container.py", line 164, in get_container
    image=get_registerable_container_image(self.container_image, settings.image_config),
  File "/Users/kevin/git/flytekit/flytekit/core/python_auto_container.py", line 235, in get_registerable_container_image
    raise AssertionError(f"Image Config with name {name} not found in the configuration")
AssertionError: Image Config with name trainer not found in the configuration
```

## Type
 - [x] Bug Fix
 - [ ] Feature
 - [ ] Plugin

## Are all requirements met?

 - [ ] Code completed
 - [ ] Smoke tested
 - [ ] Unit tests added
 - [ ] Code documentation added
 - [ ] Any pending items have an associated Issue

## Complete description
 _How did you fix the bug, make the feature etc. Link to any design docs etc_

## Tracking Issue
https://github.com/flyteorg/flyte/issues/<number>

## Follow-up issue
_NA_
OR
_https://github.com/flyteorg/flyte/issues/<number>_

### CODE REVIEW

**Review by eapolinario — COMMENTED** [review]
This is not quite right yet. IIUC, this PR makes it so that the `--config` flag (which applies at the top level) becomes a required parameter. 

Also, the `images` config section has no equivalent in the yaml config, right?

**Comment by pingsutw** [comment]
sorry, updated it.
- Now `--config` flag should be optional.
- Support reading image config in yaml config

File: flytekit/clis/sdk_in_container/run.py [hunk 1]
```diff
@@ -512,12 +512,19 @@ def _run(*args, **kwargs):
             return
 
         remote = ctx.obj[FLYTE_REMOTE_INSTANCE_KEY]
+        config_file = ctx.obj.get(CTX_CONFIG_FILE)
```
Comment (eapolinario) [inline flytekit/clis/sdk_in_container/run.py:515]: just for own understanding, what's the level that config is applied? As in, can you pass the `--config` flag to the `run` command?

**Review by eapolinario — DISMISSED** [review]
LGTM.

File: flytekit/clis/sdk_in_container/run.py (hunk 1 above)
Comment (pingsutw) [reply flytekit/clis/sdk_in_container/run.py:515]: We still have to use `--config` before the `run`. like `pyflyte --config flyte.yaml run ....`

**Comment by pingsutw** [comment]
Convert it to draft, need further discussion with Yee

File: tests/flytekit/unit/configuration/configs/sample.yaml [hunk 2]
```diff
@@ -11,3 +11,6 @@ storage:
     access-key: minio
     endpoint: http://localhost:30084
     secret-key: miniostorage
+images:
```
Comment (wild-endeavor) [inline tests/flytekit/unit/configuration/configs/sample.yaml:14]: I don't think this is supported at all on the flytectl side (and might never be since it's mostly responsible for registration time settings) but that's ok.  We can still support this.

File: flytekit/configuration/internal.py [hunk 3]
```diff
@@ -19,16 +19,20 @@ def get_specified_images(cfg: ConfigFile) -> typing.Dict[str, str]:
...
+            try:
+                image_names = cfg.legacy_config.options("images")
+            except configparser.NoSectionError:
+                return {}
+            if image_names:
+                for i in image_names:
+                    images[str(i)] = cfg.legacy_config.get("images", i)
+            return images
+        if cfg.yaml_config:
+            return cfg.yaml_config.get("images")
```
Comment (wild-endeavor) [inline flytekit/configuration/internal.py:35]: this needs to return {} in case it doesn't exist.

File: flytekit/configuration/__init__.py [hunk 4]
```diff
@@ -237,18 +237,13 @@ def auto(cls, config_file: typing.Union[str, ConfigFile] = None, img_name: Optio
         :param img_name:
         :return:
         """
-        if config_file is None and img_name is None:
-            raise ValueError("Either an image or a config with a default image should be provided")
```
Comment (wild-endeavor) [inline flytekit/configuration/__init__.py:241]: I think we need to keep this right?  Why get rid of this?  What should happen if both are None?

File: flytekit/clis/sdk_in_container/run.py [hunk 5]
```diff
@@ -512,12 +512,19 @@ def _run(*args, **kwargs):
...
+
+        image_config = run_level_params.get("image_config", None)
+        if image_config:
+            image_config.images.extend(ImageConfig.auto(config_file).images)
```
Comment (wild-endeavor) [inline flytekit/clis/sdk_in_container/run.py:519]: also how does this work? weird.  isn't ImageConfig supposed to be frozen??

File: flytekit/configuration/__init__.py (hunk 4 above)
Comment (pingsutw) [reply flytekit/configuration/__init__.py:241]: If config file is None, we should use [get_config_file](https://github.com/flyteorg/flytekit/blob/195d26ab0cc03260bb96b828219fcf7e6ea768a1/flytekit/configuration/file.py#L231) to read config from default path or from env variables

Comment (wild-endeavor) [reply flytekit/configuration/__init__.py:241]: got it yeah, it was returning None earlier, fixed that.

**Review by wild-endeavor — APPROVED** [review]
